Mathlib/Data/Matrix/Basic.lean

```diff
@@ -0,0 +1,2823 @@
+def row (v : n → α) : Matrix Unit n α :=
+  of fun _ y => v y
+#align matrix.row Matrix.row
+
+-- TODO: set as an equation lemma for `row`, see mathlib4#3024
+@[simp]
+theorem row_apply (v : n → α) (i j) : row v i j = v j :=
+  rfl
+#align matrix.row_apply Matrix.row_apply
+
+instance inhabited [Inhabited α] : Inhabited (Matrix m n α) :=
+  instInhabitedForAll_1 _
+-- Porting note: this instance was called `Pi.instance` in lean3-core, which is much
+-- nicer than the name `instInhabitedForAll_1` it got in lean4-core...
+
+instance add [Add α] : Add (Matrix m n α) :=
+  Pi.instAdd
+
+instance addSemigroup [AddSemigroup α] : AddSemigroup (Matrix m n α) :=
+  Pi.addSemigroup
+
+instance addCommSemigroup [AddCommSemigroup α] : AddCommSemigroup (Matrix m n α) :=
+  Pi.addCommSemigroup
+
+instance zero [Zero α] : Zero (Matrix m n α) :=
+  Pi.instZero
+
+instance addZeroClass [AddZeroClass α] : AddZeroClass (Matrix m n α) :=
+  Pi.addZeroClass
+
+instance addMonoid [AddMonoid α] : AddMonoid (Matrix m n α) :=
+  Pi.addMonoid
+
+instance addCommMonoid [AddCommMonoid α] : AddCommMonoid (Matrix m n α) :=
+  Pi.addCommMonoid
+
+instance neg [Neg α] : Neg (Matrix m n α) :=
+  Pi.instNeg
+
+instance sub [Sub α] : Sub (Matrix m n α) :=
+  Pi.instSub
+
+instance addGroup [AddGroup α] : AddGroup (Matrix m n α) :=
+  Pi.addGroup
+
+instance addCommGroup [AddCommGroup α] : AddCommGroup (Matrix m n α) :=
+  Pi.addCommGroup
+
+instance unique [Unique α] : Unique (Matrix m n α) :=
+  Pi.unique
+
+instance subsingleton [Subsingleton α] : Subsingleton (Matrix m n α) :=
+  instSubsingletonForAll
+--Porting note: this instance was `Pi.subsingleton` in lean3-core
+
+instance nonempty [Nonempty m] [Nonempty n] [Nontrivial α] : Nontrivial (Matrix m n α) :=
+  Function.nontrivial
+
+instance smul [SMul R α] : SMul R (Matrix m n α) :=
+  Pi.instSMul
+
+instance smulCommClass [SMul R α] [SMul S α] [SMulCommClass R S α] :
+    SMulCommClass R S (Matrix m n α) :=
+  Pi.smulCommClass
+
+instance isScalarTower [SMul R S] [SMul R α] [SMul S α] [IsScalarTower R S α] :
+    IsScalarTower R S (Matrix m n α) :=
+  Pi.isScalarTower
+
+instance isCentralScalar [SMul R α] [SMul Rᵐᵒᵖ α] [IsCentralScalar R α] :
+    IsCentralScalar R (Matrix m n α) :=
+  Pi.isCentralScalar
+
+instance mulAction [Monoid R] [MulAction R α] : MulAction R (Matrix m n α) :=
+  Pi.mulAction _
+
+instance distribMulAction [Monoid R] [AddMonoid α] [DistribMulAction R α] :
+    DistribMulAction R (Matrix m n α) :=
+  Pi.distribMulAction _
+
+instance module [Semiring R] [AddCommMonoid α] [Module R α] : Module R (Matrix m n α) :=
+  Pi.module _ _ _
+
+-- Porting note: added this section with simp lemmas because `simp` fails
+-- to apply the corresponding lemmas in the namespace `Pi`.
+-- (e.g. `Pi.zero_apply` used on `OfNat.ofNat 0 i j`)
+section NewSimpLemmas
+
+@[simp]
+theorem zero_apply [Zero α] (i : m) (j : n) : (0 : Matrix m n α) i j = 0 := rfl
```

Comment:
I added these lemmas because `simp` could not apply the same lemmas inside the `Pi` namespace. This solved all problems with `simp`. Is this the right thing to do or does anybody have a better solution?

Reply:
I think this is probably the right thing to do.

Reply:
ok. should I remove the `section NewSimpLemmas`? or give it a better name?

Reply:
I think an un-named section is fine; it's nice to have something to scope the porting note